### PR TITLE
GODRIVER-1605 Add SetBatchSize to mongo.Cursor

### PR DESCRIPTION
**Pull request by prestonvasquez**
GODRIVER-1605

## Summary

The "SetBatchSize" method can be used to override the batch size of a mongo.Cursor instance. This allows users to customize the batch size according to their needs.

## Background & Motivation

The cursor's "Next" method executes a [getMore](https://www.mongodb.com/docs/v6.0/reference/command/getMore/) operation with a specific batch size. However, there may be cases where it is necessary to override the default batch size in order to prevent cursor operations from timing out. The Go Driver supports this use case by allowing the batch size to be overridden.

### CODE REVIEW

**Review by qingyang-hu — COMMENTED** [review]
This change works functionally. However, the API seems confusing to me with the batch size in the option. Do we need a more expressive function name other than the `SetBatchSize`?

**Comment by prestonvasquez** [comment]
> This change works functionally. However, the API seems confusing to me with the batch size in the option. Do we need a more expressive function name other than the `SetNatchSize`?

@qingyang-hu AFAIK "SetBatchSize" was added to "drivers.BatchCursor" in GODRIVER-1900, which appears to have sourced the name from [cursor.batchSize()](https://www.mongodb.com/docs/manual/reference/method/cursor.batchSize/#mongodb-method-cursor.batchSize). This PR proposed to extend this logic to the "mongo.Cursor". Here is a UML diagram:

Screenshot 2023-03-16 at 11 10 28 AM

Do you have thoughts on how to rename the method to improve clarity?

File: x/mongo/driver/batch_cursor.go [hunk 1]
```diff
@@ -308,25 +308,39 @@ func (bc *BatchCursor) KillCursor(ctx context.Context) error {
...
+	return gmBatchSize, nil
+}
+
+func (bc *BatchCursor) getMore(ctx context.Context) {
+	bc.clearBatch()
+	if bc.id == 0 {
+		return
+	}
+
+	numToReturn, err := bc.getMoreBatchSize(ctx)
```
Comment (matthewdale) [inline x/mongo/driver/batch_cursor.go:337]: This refactor changes the behavior of `getMore` if the "limit" emulation overrides the batch size to be <= 0. Previously, `getMore` would close the cursor and return. Now, it closes the cursor, sets `numToReturn` to 0, and continues to run the operation.

Comment (prestonvasquez) [reply x/mongo/driver/batch_cursor.go]: Really good catch, thank you. I've updated the refactored function to return an "ok" value. The reason for breaking out the function was principally for unit testing.

File: x/mongo/driver/batch_cursor.go [hunk 2]
```diff
@@ -308,23 +308,36 @@ func (bc *BatchCursor) KillCursor(ctx context.Context) error {
 	}.Execute(ctx)
 }
 
+// calcGetMoreBatchSize returns the number of documents that should be returned in
+// the response of a "getMore" operation given the limit, batchSize, and number
+// of documents already returned.
+func calcGetMoreBatchSize(ctx context.Context, bc BatchCursor) (int32, bool) {
```
Comment (matthewdale) [inline x/mongo/driver/batch_cursor.go:314]: The context parameter is unused.

```suggestion
func calcGetMoreBatchSize(bc BatchCursor) (int32, bool) {
```

File: x/mongo/driver/batch_cursor.go [hunk 3]
```diff
@@ -308,23 +308,36 @@ func (bc *BatchCursor) KillCursor(ctx context.Context) error {
 	}.Execute(ctx)
 }
 
+// calcGetMoreBatchSize returns the number of documents that should be returned in
+// the response of a "getMore" operation given the limit, batchSize, and number
+// of documents already returned.
```
Comment (matthewdale) [inline x/mongo/driver/batch_cursor.go:313]: Optional: Add a description of what the `bool` return value indicates.

**Review by matthewdale — APPROVED** [review]
Looks good 👍

File: x/mongo/driver/batch_cursor.go (hunk 2 above)
Comment (qingyang-hu) [reply x/mongo/driver/batch_cursor.go:314]: I didn't quite get the motivation of changing this function from a method of `BatchCursor` to an independent function.

Comment (prestonvasquez) [reply x/mongo/driver/batch_cursor.go]: The purpose is to unit test the calculation.

**Review by qingyang-hu — APPROVED** [review]
LGTM!

**Pull request merged**